### ISSUE TITLE
Build errors when using Span<T> with Xamarin.Android/Mono projects

### ISSUE THREAD

**Issue by andriysavin** [dotnet/corefx#29654]
System.Memory package claims to support Xamarin.Android as well as `netstandard`. However, all is not that straight. Consider the following scenario:  in a solution there is a Xamarin.Android project, which references a `netstandard2.0` class library project. The class library references System.Memory package, and exposes a type which has `Span<T>` in method signatures. When I'm building such a solution, I'm getting various errors depending on how I consume the type from the class library, e.g:
```
> ...\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(1590,2): error : Exception while loading assemblies: System.IO.FileNotFoundException: Could not load assembly 'System.Memory, Version=4.0.1.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'. Perhaps it doesn't exist in the Mono for Android profile?
> ...\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(1590,2): error : File name: 'System.Memory.dll'
> ...\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(1590,2): error :    at Java.Interop.Tools.Cecil.DirectoryAssemblyResolver.Resolve(AssemblyNameReference reference, ReaderParameters parameters)
> ...\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(1590,2): error :    at Xamarin.Android.Tasks.ResolveAssemblies.AddAssemblyReferences(DirectoryAssemblyResolver resolver, ICollection`1 assemblies, AssemblyDefinition assembly, Boolean topLevel)
> ...\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(1590,2): error :    at Xamarin.Android.Tasks.ResolveAssemblies.AddAssemblyReferences(DirectoryAssemblyResolver resolver, ICollection`1 assemblies, AssemblyDefinition assembly, Boolean topLevel)
> ...\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(1590,2): error :    at Xamarin.Android.Tasks.ResolveAssemblies.Execute(DirectoryAssemblyResolver resolver)
> 

> ...\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(1590,2): warning : Failed to read 'C:\Users\andsav\.nuget\packages\system.memory\4.5.0-rc1\lib\MonoAndroid10\_._' with debugging symbols.Retrying to load it without it.Error details are logged below.
> ...\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(1590,2): warning : System.BadImageFormatException: Format of the executable(.exe) or library(.dll) is invalid.
> ...\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(1590,2): warning :    at Mono.Cecil.PE.ImageReader.ReadImage()
> ...\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(1590,2): warning :    at Mono.Cecil.PE.ImageReader.ReadImage(Disposable`1 stream, String file_name)
> ...\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(1590,2): warning :    at Mono.Cecil.ModuleDefinition.ReadModule(String fileName, ReaderParameters parameters)
> ...\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(1590,2): warning :    at Java.Interop.Tools.Cecil.DirectoryAssemblyResolver.ReadAssembly(String file)
> ...\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(1590,2): error : Exception while loading assemblies: Java.Interop.Tools.Diagnostics.XamarinAndroidException: error XA0009: Error while loading assembly: C:\Users\andsav\.nuget\packages\system.memory\4.5.0-rc1\lib\MonoAndroid10\_._ ---> System.BadImageFormatException: Format of the executable(.exe) or library(.dll) is invalid.
> ...\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(1590,2): error :    at Mono.Cecil.PE.ImageReader.ReadImage()
> ...\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(1590,2): error :    at Mono.Cecil.PE.ImageReader.ReadImage(Disposable`1 stream, String file_name)
> ...\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(1590,2): error :    at Mono.Cecil.ModuleDefinition.ReadModule(String fileName, ReaderParameters parameters)
> ...\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(1590,2): error :    at Java.Interop.Tools.Cecil.DirectoryAssemblyResolver.ReadAssembly(String file)
> ...\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(1590,2): error :    at Java.Interop.Tools.Cecil.DirectoryAssemblyResolver.Load(String fileName, Boolean forceLoad)
> ...\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(1590,2): error :    --- End of inner exception stack trace ---
> ...\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(1590,2): error :    at Java.Interop.Tools.Cecil.DirectoryAssemblyResolver.Load(String fileName, Boolean forceLoad)
> ...\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(1590,2): error :    at Xamarin.Android.Tasks.ResolveAssemblies.ResolveRuntimeAssemblyForReferenceAssembly(LockFile lockFile, DirectoryAssemblyResolver resolver, String assemblyPath)

```
`>error CS0433: The type 'Span<T>' exists in both 'System.Memory, Version=4.0.1.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' and 'mscorlib, Version=2.0.5.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e'`

I understand that this repo is likely not the right place to file this issue, but I'm not sure if the problem is related to how the package is built or to how Mono implements `Span<T>` and friends. The end result is that this package can't be consumed in all scenarios it claims to support, and `netstandard` libraries consuming this package basically stop to work on one of `nestandard` target platforms.

[edited by @danmosemsft to add ticks for formatting]

**Comment by danmosemsft** [comment]
@weshaggard?

**Comment by karelz** [comment]
cc @marek-safar

**Comment by weshaggard** [comment]
Yes I do believe it is the same thing and needs to be fixed on the Xamarin side of things.

**Comment by joperezr** [comment]
> is the root cause of the issue the same as #29523?

Yup same thing as @weshaggard suggests.

**Comment by andriysavin** [comment]
Thanks for explaining! Maybe someone can point me where at Xamarin (Mono?) should this issue be filed then?

**Comment by marek-safar** [comment]
What Xamarin version are you using? Unless it's very recent preview version it won't work. This is a preview version of the nuget which requires preview version of future Xamarin release.

**Comment by marek-safar** [comment]
You could see that from About Visual Studio. I checked it and Android version is still missing System.Memory facade only iOS works. I was testing with 8.3 (XA) and 11.11 (XI).

**Comment by andriysavin** [comment]
> Xamarin   4.10.0.442 (396b18cef)
> Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.
> 
> Xamarin Designer   4.12.264 (fc37cd02e)
> Visual Studio extension to enable Xamarin Designer tools in Visual Studio.
> 
> Xamarin.Android SDK   8.3.0.19 (HEAD/342b2ce96)
> Xamarin.Android Reference Assemblies and MSBuild support.
> 
> Xamarin.iOS and Xamarin.Mac SDK   11.10.1.177 (7e782c1)
> Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.

**Comment by marek-safar** [comment]
This is now tracked as #29847 as well as updated to Xamarin.Android in 8.4 version

**Comment by danmosemsft** [comment]
Can this be closed as dupe of #29847

**Comment by mohakbarokar** [comment]
This Bug is tested on build with config: https://gist.github.com/mohakbarokar/2a6889577bc362531d623eae48c5b211

Ide Build taken from: \\cpvsbuild\drops\VS\d15.8\products\27729.01

Xamarin.Android Taken from: https://jenkins.mono-project.com/view/Xamarin.Android/job/xamarin-android/lastSuccessfulBuild/Azure/

Nuget Package Taken from: \\TAREKMS-DEV\PrivatePackages

Steps followed for verification:
1. Creates single view android project
2. Added System.Memory Nuget package 
3. Added code
 private static Span<char> Trim(Span<char> source)
       {
           var arr = new byte[10];
           Span<byte> bytes = arr;

           return source;
       }

4. Built project

We observed that we got resolve conflicts for _System.Runtime.CompilerServices.Unsafe_ & _System.Buffers_ after adding those packages build was successful, We discussed this with @marek-safar  and he suggested that this is expected behaviour.

As the error mentioned in the bug is not reproducible and the behavior is expected we can mark the bug as Fixed.

Screencast Link: 
1. https://www.screencast.com/t/p1zA85cLhFa
2. https://www.screencast.com/t/YVaLy657v
